examples/ecs/indexing.rs

```diff
@@ -0,0 +1,205 @@
+
+use bevy::{
+    ecs::{
+        component::Tick,
+        system::{SystemChangeTick, SystemParam},
+    },
+    prelude::*,
+    utils::{HashMap, HashSet},
+};
+use bevy_internal::ecs::query::ReadOnlyWorldQuery;
+use std::{hash::Hash, marker::PhantomData};
+
+pub trait Indexer {
+    type Input: Component;
+    type Index: Hash + Eq + Clone + 'static;
+
+    fn index(input: &Self::Input) -> Self::Index;
+}
+
+pub struct SimpleIndexer<T>(PhantomData<T>);
+
+impl<T> Indexer for SimpleIndexer<T> where T: Component + Hash + Eq + Clone {
+    type Input = T;
+
+    type Index = T;
+
+    fn index(input: &Self::Input) -> Self::Index {
+        input.clone()
+    }
+}
+
+#[derive(Resource)]
+struct IndexBacking<T, F = (), I = SimpleIndexer<T>> {
+    forward: HashMap<T, HashSet<Entity>>,
+    reverse: HashMap<Entity, T>,
+    last_this_run: Option<Tick>,
+    _phantom: PhantomData<fn(F, I)>,
+}
+
+impl<T, F, I> Default for IndexBacking<T, F, I> {
+    fn default() -> Self {
+        Self {
+            forward: default(),
+            reverse: default(),
+            last_this_run: default(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T, F> IndexBacking<T, F> {
+    fn update(&mut self, entity: Entity, value: Option<T>) -> Option<T>
+    where
+        T: Hash + Eq + Clone,
+    {
+        let old = if let Some(ref value) = value {
+            self.reverse.insert(entity, value.clone())
+        } else {
+            self.reverse.remove(&entity)
+        };
+
+        if let Some(ref old) = old {
+            if let Some(set) = self.forward.get_mut(old) {
+                set.remove(&entity);
+
+                if set.is_empty() {
+                    self.forward.remove(old);
+                }
+            }
+        }
+
+        if let Some(value) = value {
+            self.forward.entry(value).or_default().insert(entity);
+        };
+
+        old
+    }
+
+    fn get(&self, value: &T) -> Option<impl Iterator<Item = Entity> + '_>
+    where
+        T: Hash + Eq + Clone,
+    {
+        Some(self.forward.get(value)?.iter().copied())
+    }
+}
+
+#[derive(SystemParam)]
+pub struct Index<'w, 's, T, F = ()>
+where
+    T: Component + Hash + Eq,
+    F: ReadOnlyWorldQuery + 'static,
+{
+    changed: Query<'w, 's, (Entity, Ref<'static, T>), (Changed<T>, F)>,
+    removed: RemovedComponents<'w, 's, T>,
+    index: ResMut<'w, IndexBacking<T, F>>,
+    this_run: SystemChangeTick,
+}
+
+impl<'w, 's, T, F> Index<'w, 's, T, F>
+where
+    T: Component + Hash + Eq + Clone,
+    F: ReadOnlyWorldQuery + 'static,
+{
+    fn update_index_internal(&mut self) {
+        let this_run = self.this_run.this_run();
+
+        // Remove old entires
+        for entity in self.removed.read() {
+            self.index.update(entity, None);
+        }
+
+        // Update new and existing entries
+        for (entity, component) in self.changed.iter() {
+            self.index.update(entity, Some(component.clone()));
+        }
```

Comment:
This will require iterating over all entities that have a `T` component every time a `Index` may be out of sync, which IMO isn't ideal. AFAIK `bevy_mod_index` explored this possibility and it wasn't very efficient.

Reply:
isnt the self.changed query specifically filtering only changed entities?